### PR TITLE
[AMD] [diffusion] feat: enable AITer GroupNorm for VAE decode on ROCm

### PR DESCRIPTION
**Pull request by yctseng0211**
## Summary
Replace `nn.GroupNorm` with aiter's CK  GroupNorm in VAE decode stage on ROCm, controlled by `SGLANG_USE_ROCM_VAE=1` (off by default).

## Motivation
Profiling Z-Image-Turbo VAE decode on MI300X/MI35X shows PyTorch's native `RowwiseMomentsCUDAKernel` is poorly optimized for AMD GPUs, 1.57ms per call vs AITer's 0.13ms (12x faster).

## Results (Z-Image-Turbo, 1024×1024, MI300X)
max_concurrency=8, num_prompts=64

| Metric | Baseline | AITer GroupNorm | Change |
|--------|----------|-----------------|--------|
| GroupNorm kernel time | 51.4ms | 6.0ms | **-88%** |
| VAE decode total | ~205ms | ~162ms | **-21%** |
| End-to-end throughput | 0.745 | 0.770 | **+3~5%** |

## Modifications
- `decoding.py`: recursively swap `nn.GroupNorm` → `aiter.GroupNorm` after VAE load
- Zero-copy weight sharing; safe fallback on error

### Design note
GroupNorm lives in diffusers' `Decoder` / `ResnetBlock2D` / `UNetMidBlock2D`, not in SGLang code.
We use runtime module replacement (`named_children` + `setattr`) rather than patching diffusers directly so that:
- The change is opt-in (`SGLANG_USE_ROCM_VAE=1`) and has no effect on non-ROCm platforms
- Diffusers upgrades won't break or overwrite the optimization
- Only the VAE decode stage is affected; other pipelines remain untouched

## Accuracy Tests

```
export SGLANG_USE_ROCM_VAE=1
sglang generate --model-path=Tongyi-MAI/Z-Image-Turbo --log-level=info --prompt='A fantasy landscape with mountains and a river, detailed, vibrant colors' --width=1024 --height=1024 --num-inference-steps=9 --guidance-scale=0.0 --seed=42 --save-output --enable-torch-compile --warmup --dit-cpu-offload false --text-encoder-cpu-offload false
```

image

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

**Review by HaiShaw — COMMENTED** [review]
Suggest to change `SGLANG_USE_AITER_VAE` to `SGLANG_USE_ROCM_VAE`, which/env can be reused for future ROCm based VAE optimizations.

**Review by HaiShaw — APPROVED** [review]
LGTM

**Review by yichiche — APPROVED** [review]
LGTM.

**Review by michaelzhang-ai — APPROVED** [review]

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/decoding.py [hunk 1]
```diff
@@ -22,11 +22,42 @@
...
 from sglang.multimodal_gen.runtime.platforms import current_platform
 from sglang.multimodal_gen.runtime.server_args import ServerArgs, get_global_server_args
+from sglang.multimodal_gen.runtime.utils.common import get_bool_env_var
 from sglang.multimodal_gen.runtime.utils.logging_utils import init_logger
 from sglang.multimodal_gen.utils import PRECISION_TO_TYPE
 
 logger = init_logger(__name__)
 
+_is_hip = current_platform.is_hip()
+_use_aiter_vae = get_bool_env_var("SGLANG_USE_ROCM_VAE") and _is_hip
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/decoding.py:32]: please register them via `envs.py`

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/decoding.py [hunk 2]
```diff
@@ -22,11 +22,42 @@
...
+
+
+def _replace_groupnorm_with_aiter(module: torch.nn.Module) -> int:
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/decoding.py:35]: These patterns (including `_is_hip`) should be avoided.
Decoding is a critical core stage of the system. If we allow platform-specific conditionals to accumulate here, the complexity will grow rapidly and become increasingly hard to manage.

Over time, this will significantly hurt maintainability and extensibility.

Suggestions:

* Introduce a Platform abstraction
* Consider a plugin-based architecture

discussions are welcomed

Comment (mickqian) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/decoding.py:35]: The codebase already has a well-established pattern for isolating platform-specific logic via `Platform` polymorphic methods:

- `Platform.get_attn_backend_cls_str()` — platform decides which attention backend to use
- `Platform.enable_dit_layerwise_offload_for_wan_by_default()` — platform decides default behavior
- `Platform.get_device_communicator_cls()` — platform decides communicator

We can directly reuse this Strategy pattern:

### 1. Add `optimize_vae()` hook to `Platform` base class

```python
# interface.py
class Platform:
    ...
    @classmethod
    def optimize_vae(cls, vae: torch.nn.Module) -> torch.nn.Module:
        “”"Apply platform-specific optimizations to VAE after loading.“”"
        return vae
```

### 2. `RocmPlatform` overrides it with AITer GroupNorm replacement

```python
# rocm.py
class RocmPlatform(Platform):
    ...
    @classmethod
    def optimize_vae(cls, vae: torch.nn.Module) -> torch.nn.Module:
        if not envs.SGLANG_USE_ROCM_VAE:
            return vae
        try:
            from aiter.ops.groupnorm import GroupNorm as AITerGroupNorm
            count = cls._replace_groupnorm(vae, AITerGroupNorm)
            logger.info(“replaced %d GroupNorm with AITer GroupNorm”, count)
        except Exception as e:
            logger.warning(“failed to apply AITer GroupNorm: %s”, e)
        return vae

    @staticmethod
    def _replace_groupnorm(module, aiter_gn_cls):
        count = 0
        for name, child in module.named_children():
            if isinstance(child, torch.nn.GroupNorm):
                replacement = aiter_gn_cls(child.num_groups, child.num_channels, ...)
                replacement.weight = child.weight  # zero-copy weight sharing
                replacement.bias = child.bias
                setattr(module, name, replacement)
                count += 1
            else:
                count += RocmPlatform._replace_groupnorm(child, aiter_gn_cls)
        return count
```

### 3. Call it from `VAELoader` — one line

`VAELoader` already does similar model-level optimizations (e.g. `_convert_conv3d_weights_to_channels_last_3d`). Just add one line before returning:

```python
# vae_loader.py
vae = current_platform.optimize_vae(vae)
return vae
```

### 4. `decoding.py` — zero changes needed

The core decoding stage stays completely platform-agnostic. No `_is_hip`, no `_use_aiter_vae`, no `_replace_groupnorm_with_aiter`.

**Why this is the right approach:**
- Follows the existing architecture exactly (same pattern as `get_attn_backend_cls_str`)
- Separation of concerns: ROCm code lives only in `rocm.py`; core stages don’t know about platforms
- Extensible: future NPU/MUSA VAE optimizations just override `optimize_vae()` in their own platform file — no core code changes needed
- Env var `SGLANG_USE_ROCM_VAE` is registered in `envs.py` and consumed internally by `RocmPlatform`
- Minimal diff: ~3 files, a few lines each

Comment (yctseng0211) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/decoding.py:35]: @mickqian Thanks for the detailed review and suggestions, I'll follow the Platform abstraction pattern and address this in a follow-up PR.

Comment (yctseng0211) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/decoding.py:35]: Hi @mickqian could you help review this PR-20496 https://github.com/sgl-project/sglang/pull/20496? 
This PR moves platform-specific AITer GroupNorm replacement out of decoding.py into the Platform polymorphic pattern:

- Register SGLANG_USE_ROCM_VAE in envs.py
- Add optimize_vae() hook to Platform base class
- Override in RocmPlatform with AITer GroupNorm replacement
- Call from VAELoader at load time
- Remove all platform-specific logic from decoding.py

Thanks.
